src/function/gds/strongly_connected_components.cpp

```diff
@@ -0,0 +1,245 @@
+#include <iostream>
+#include <thread>
+#include <tuple>
+#include <vector>
+
+#include "binder/binder.h"
+#include "binder/expression/expression_util.h"
+#include "common/types/types.h"
+#include "function/gds/auxiliary_state/gds_auxilary_state.h"
+#include "function/gds/gds_frontier.h"
+#include "function/gds/gds_function_collection.h"
+#include "function/gds/gds_object_manager.h"
+#include "function/gds/gds_utils.h"
+#include "function/gds/output_writer.h"
+#include "function/gds_function.h"
+#include "graph/graph.h"
+#include "processor/execution_context.h"
+#include "processor/result/factorized_table.h"
+#include <format>
+
+using namespace std;
+using namespace kuzu::binder;
+using namespace kuzu::common;
+using namespace kuzu::processor;
+using namespace kuzu::storage;
+using namespace kuzu::graph;
+
+namespace kuzu {
+namespace function {
+
+offset_t NOT_ASSIGNED = std::numeric_limits<offset_t>::max();
+
+/// State for the SCC algorithm
+class SCCState {
+public:
+    SCCState(const table_id_map_t<offset_t>& numNodesMap, MemoryManager* mm) {
+        for (const auto& [tableID, numNodes] : numNodesMap) {
+            componentIDsMap.allocate(tableID, numNodes, mm);
+            visitsMap.allocate(tableID, numNodes, mm);
+            pinTable(tableID);
+            for (auto i = 0u; i < numNodes; ++i) {
+                visits[i].store(false, std::memory_order_relaxed);
+                componentIDs[i].store(NOT_ASSIGNED, std::memory_order_relaxed);
+            }
+        }
+    }
+
+    void pinTable(table_id_t tableID) {
+        componentIDs = componentIDsMap.getData(tableID);
+        visits = visitsMap.getData(tableID);
+    }
+
+    bool updateVisit(common::offset_t offset, bool value) {
+        visits[offset].store(value, std::memory_order_relaxed);
+        return true;
+    }
+
+    bool getVisit(offset_t offset) { return visits[offset].load(std::memory_order_relaxed); }
+
+    bool updateComponentID(common::offset_t offset, offset_t value) {
+        componentIDs[offset].store(value, std::memory_order_relaxed);
+        return true;
+    }
+
+    offset_t getComponentID(offset_t offset) {
+        return componentIDs[offset].load(std::memory_order_relaxed);
+    }
+
+private:
+    // Pointers to current tableID's visits and componentIDs
+    std::atomic<bool>* visits = nullptr;
+    std::atomic<offset_t>* componentIDs = nullptr;
```

Comment:
If I remove the `atomic`s, I get a CTE: `error: member reference base type 'bool' is not a structure or union`. How do I use non-atomic vars here?

Reply:
Thanks, yeah I might have missed one of them.